tools/code_analyzer/op_dependency.cpp

```diff
@@ -0,0 +1,396 @@
+
+    if (OutputFormat == OutputFormatType::Dot) {
+      printAsDot(std::cout, keyNodes, result);
+    } else if (OutputFormat == OutputFormatType::YAML) {
+      printAsYAML(std::cout, keyNodes, result);
+    }
+
+    return false;
+  }
+
+private:
+  // Searching for global string constants that look like function schema, e.g.:
+  // "aten::relu". Then search for places using the function schema string
+  // constant to 1) register op; 2) invoke op.
+  static void scanGlobalsForFunctionSchema(
+      Module& M, SET* keyNodes,
+      GRAPH* schemaStrToFunctions, GRAPH* functionsToSchemaStrs) {
+    for (GlobalVariable& GV : M.globals()) {
+      if (!GV.hasInitializer()) continue;
+      Constant* gvInit = GV.getInitializer();
+      if (gvInit->isNullValue()) continue;
+      ConstantDataArray* gvInitArray = dyn_cast<ConstantDataArray>(gvInit);
+      if (!gvInitArray || !gvInitArray->isCString()) continue;
+      std::string gvInitStr = gvInitArray->getAsCString().str();
+      if (!isFunctionSchemaString(gvInitStr)) continue;
+      std::string schemaStr = truncateFunctionSchemaString(gvInitStr);
+
+      // Track all seen schema strings. Useful for checking whether we miss any
+      // dependency in the logic below.
+      keyNodes->insert(schemaStr);
+
+      // Search for GV users.
+      for (auto gvUser : GV.users()) {
+        // Search for "i8* getelementptr inbounds" ConstantExpr.
+        ConstantExpr* CE = dyn_cast<ConstantExpr>(gvUser);
+        if (!CE) continue;
+
+        // Search for CE users.
+        for (auto ceUser : CE->users()) {
+          // Search for "std::basic_string" constructor.
+          Instruction* ceUserInst = dyn_cast<Instruction>(ceUser);
+          if (!ceUserInst) continue;
+          analyzeFunctionSchemaUser(
+              schemaStr, *ceUserInst, schemaStrToFunctions,
+              functionsToSchemaStrs);
+        }
+      }
+    }
+  }
+
+  // Use fuzzy match to find all llvm::Function instances related to the
+  // function schema string. There are two types of relationship:
+  // 1) function schema to function (op registration);
+  // 2) function to function schema (op invocation);
+  //
+  // For op invocation, simply check whether the parent function calls
+  // 'c10::Dispatcher::findSchema' since the use of the schema string.
+  //
+  // For op registration, collect all function references between two following
+  // "c10::RegisterOperators::Options::schema" calls. The function references
+  // can be direct function call, function pointer or template argument.
+  //
+  // Note that these rules highly depend on how LLVM emit IRs. If it changes
+  // the sequence of schema string reference and c10::RegisterOperators API call
+  // then we might need use more sophisticated alias analysis, analyze clang-AST
+  // instead or simplify leverage the connection between function schema string
+  // and related function name (as they are generated by script).
```

Comment:
I understand that there's a test for the overall pipeline earlier in this test suite, but it might also be helpful to have a specific unit test for this aspect of LLVM's IR behavior, so that if it changes, it's easier to debug the issue.

Reply:
So, I think what this comment is saying here, is you just *assume* that the next instruction in the sequence of instructions is the one using the string, but you don't actually verify it actually uses the string in question XD. Yeah, that seems uber dodgy. Doesn't llvm maintain accurate use-defs information? I feel like it shouldn't be too hard to tell that the usage actually is using the string in question.

Reply:
The assumption is not that the "immediate" next instruction is the one using the string. I meant to say the:
```
string reference A / schema A / function ptr A / reference B / schema B / function ptr B / ...
```
sequence is not reordered to:
```
string ref A / string ref B / string ref C / schema A / function ptr A / schema B / function ptr B / ...
```
As long as it's interlaced as in the source code this heuristic search should be good as it searches for ALL instructions that reference to ANY function type operand between two consecutive schema() function call.